src/validation.cpp

```diff
@@ -378,6 +378,8 @@ void CChainState::MaybeUpdateMempoolForReorg(
                 }
             }
         }
+        // CheckSequenceLocks updates lp. Update the mempool entry LockPoints.
+        if (!validLP) m_mempool->mapTx.modify(it, update_lock_points(lp));
```

Comment:
I was thinking "what if we skip those transactions with `should_remove=true` here?". I guess, we'll have the same original problem: they won't be cleared because we would mistakenly think they have valid `lp` in `removeForReorg`?

Perhaps this deserves a comment extension, saying that txs to removed must be updated here as well.

Reply:
No, they'll still be cleared - `removeForReorg` just applies `check_final_and_mature` on all entries and removes them if it returns true, it doesn't check whether the `lp` is valid. I'm not 100% sure if we could skip the `should_remove=true` ones, but I think it would be fine. The `CheckSequenceLocks` function doesn't guarantee `lp` to be reliable if it returns false, and the function doesn't even get executed if `CheckFinalTx` returns false.

Reply:
I agree that it is safe to skip if the tx will be removed anyway. When this was introduced in commit 14d6324a248df50cb79fbeb5b60a978687a3b64e it updates unconditionally.

Reply:
I was just about to post the same optimization suggestion: change this to `if (!validLP && !should_remove)`.

> I guess, we'll have the same original problem: they won't be cleared because we would mistakenly think they have valid lp in removeForReorg

I don't see it that way - the tx will be removed if this lambda `check_final_and_mature()` returns `true`, so why bother updating that tx data? (I see no reason).

Reply:
Sleeping over this, now I think that it is better to leave it as is - just `if (!validLP)` because skipping update for the to-be-removed transactions will spare some CPU cycles but will increase the coupling between `check_final_and_mature()` and `removeForReorg()` - the former will assume the later will remove the transactions and the later must be aware that some transactions have invalid lock points in case it decides to do something else before removal. Plus more complex code. Not worth it.

Reply:
> skipping update for the to-be-removed transactions will spare some CPU cycles but will increase the coupling between check_final_and_mature() and removeForReorg()

Agreed 👍